common/clock/time_source.go

```diff
@@ -140,3 +143,20 @@ func (c *fakeClock) NewTimer(d time.Duration) Timer {
 func (c *fakeClock) AfterFunc(d time.Duration, f func()) Timer {
 	return c.FakeClock.AfterFunc(d, f)
 }
+
+func WithTimeout(ctx context.Context, ts TimeSource, timeout time.Duration) (context.Context, context.CancelFunc) {
+	return WithDeadline(ctx, ts, ts.Now().Add(timeout))
+}
+
+func WithDeadline(ctx context.Context, ts TimeSource, t time.Time) (context.Context, context.CancelFunc) {
```

Comment:
How about making these functions a method of the clock interface?

Reply:
Good point. I've merged it already in a different PR, but can change it :)

Reply:
Fixed in https://github.com/cadence-workflow/cadence/pull/6886